neo/Network/P2P/Peer.cs

```diff
@@ -0,0 +1,219 @@
+                    break;
+                case Timer _:
+                    OnTimer();
+                    break;
+                case Peers peers:
+                    AddPeers(peers.EndPoints);
+                    break;
+                case Connect connect:
+                    ConnectToPeer(connect.EndPoint);
+                    break;
+                case WsConnected ws:
+                    OnWsConnected(ws.Socket, ws.Remote, ws.Local);
+                    break;
+                case Tcp.Connected connected:
+                    OnTcpConnected(((IPEndPoint)connected.RemoteAddress).Unmap(), ((IPEndPoint)connected.LocalAddress).Unmap());
+                    break;
+                case Tcp.Bound _:
+                    tcp_listener = Sender;
+                    break;
+                case Tcp.CommandFailed _:
+                    break;
+                case Terminated terminated:
+                    OnTerminated(terminated.ActorRef);
+                    break;
+            }
+        }
+
+        private void OnStart(int port, int ws_port)
+        {
+            ListenerPort = port;
+            timer = Context.System.Scheduler.ScheduleTellRepeatedlyCancelable(0, 5000, Context.Self, new Timer(), ActorRefs.NoSender);
+            if ((port > 0 || ws_port > 0)
+                && localAddresses.All(p => !p.IsIPv4MappedToIPv6 || IsIntranetAddress(p))
+                && UPnP.Discover())
+            {
+                try
+                {
+                    localAddresses.Add(UPnP.GetExternalIP());
+                    if (port > 0)
+                        UPnP.ForwardPort(port, ProtocolType.Tcp, "NEO");
+                    if (ws_port > 0)
+                        UPnP.ForwardPort(ws_port, ProtocolType.Tcp, "NEO WebSocket");
+                }
+                catch { }
+            }
+            if (port > 0)
+            {
+                tcp_manager.Tell(new Tcp.Bind(Self, new IPEndPoint(IPAddress.Any, port), options: new[] { new Inet.SO.ReuseAddress(true) }));
+            }
+            if (ws_port > 0)
+            {
+                ws_host = new WebHostBuilder().UseKestrel().UseUrls($"http://*:{ws_port}").Configure(app => app.UseWebSockets().Run(ProcessWebSocketAsync)).Build();
+                ws_host.Start();
+            }
+        }
+
+        private void OnTcpConnected(IPEndPoint remote, IPEndPoint local)
+        {
+            ConnectedAddresses.TryGetValue(remote.Address, out int count);
+            if (count >= MaxConnectionsPerAddress)
+            {
+                Sender.Tell(Tcp.Abort.Instance);
+            }
+            else
+            {
+                ConnectedAddresses[remote.Address] = count + 1;
+                IActorRef connection = Context.ActorOf(ProtocolProps(Sender, remote, local), $"connection_{Guid.NewGuid()}");
+                Context.Watch(connection);
+                Sender.Tell(new Tcp.Register(connection));
+                ConnectedPeers.TryAdd(connection, remote);
```

Comment:
This is a good proposal. But can you create an issue for this separately? I intend to merge this PR into the master branch first, otherwise it will get bigger and bigger.

Reply:
Seems reasonable; I’ll open an issue about it later.